File: src/com/hendraanggrian/support/utils/content/Configurations.java
package com.hendraanggrian.support.utils.content;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import static com.hendraanggrian.support.utils.Preconditions.checkNotNull;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
public final class Configurations {

    @IntDef({
            Configuration.SCREENLAYOUT_SIZE_UNDEFINED,
            Configuration.SCREENLAYOUT_SIZE_SMALL,
            Configuration.SCREENLAYOUT_SIZE_NORMAL,
            Configuration.SCREENLAYOUT_SIZE_LARGE,
            Configuration.SCREENLAYOUT_SIZE_XLARGE
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ScreenSize {
    }

    @RequiresApi(20)
    @TargetApi(20)
    @IntDef({
            Configuration.UI_MODE_TYPE_UNDEFINED,
            Configuration.UI_MODE_TYPE_NORMAL,
            Configuration.UI_MODE_TYPE_DESK,
            Configuration.UI_MODE_TYPE_CAR,
            Configuration.UI_MODE_TYPE_TELEVISION,
            Configuration.UI_MODE_TYPE_APPLIANCE,
            Configuration.UI_MODE_TYPE_WATCH
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ModeType {
    }

    @IntDef({
            Configuration.UI_MODE_NIGHT_UNDEFINED,
            Configuration.UI_MODE_NIGHT_NO,
            Configuration.UI_MODE_NIGHT_YES
    })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ModeNight {
    }

    private Configurations() {
    }

    @ScreenSize
    public static int getScreenSize(@NonNull Context context) {
        checkNotNull(context);
        return getScreenSize(context.getResources());
    }

    @ScreenSize
    public static int getScreenSize(@NonNull Resources res) {
        checkNotNull(res);
        return getScreenSize(res.getConfiguration());
    }

    @ScreenSize
    public static int getScreenSize(@NonNull Configuration config) {
        checkNotNull(config);
        @ScreenSize int cur = config.screenLayout & Configuration.SCREENLAYOUT_SIZE_MASK;
        return cur;
    }

    public static boolean isScreenSizeAtLeast(@NonNull Context context, @ScreenSize int size) {
        checkNotNull(context);
        return isScreenSizeAtLeast(context.getResources(), size);
    }

    public static boolean isScreenSizeAtLeast(@NonNull Resources res, @ScreenSize int size) {
        checkNotNull(res);
        return isScreenSizeAtLeast(res.getConfiguration(), size);
    }

    public static boolean isScreenSizeAtLeast(@NonNull Configuration config, @ScreenSize int size) {
        checkNotNull(config);
        int cur = getScreenSize(config);
        return cur != Configuration.SCREENLAYOUT_SIZE_UNDEFINED && cur >= size;
    }

    public static boolean isScreenLong(@NonNull Context context) {
        checkNotNull(context);
        return isScreenLong(context.getResources());
    }

    public static boolean isScreenLong(@NonNull Resources res) {
        checkNotNull(res);
        return isScreenLong(res.getConfiguration());
    }

    public static boolean isScreenLong(@NonNull Configuration config) {
        checkNotNull(config);
        return (config.screenLayout & Configuration.SCREENLAYOUT_LONG_MASK) == Configuration.SCREENLAYOUT_LONG_YES;
    }

    @RequiresApi(23)
    @TargetApi(23)
    public static boolean isScreenRound(@NonNull Context context) {
        checkNotNull(context);
        return isScreenRound(context.getResources());
    }

    @RequiresApi(23)
    @TargetApi(23)
    public static boolean isScreenRound(@NonNull Resources res) {
        checkNotNull(res);
        return isScreenRound(res.getConfiguration());
    }

    @RequiresApi(23)
    @TargetApi(23)
    public static boolean isScreenRound(@NonNull Configuration config) {
        checkNotNull(config);
        return config.isScreenRound();
    }

    @RequiresApi(17)
    @TargetApi(17)
    public static boolean isRtl(@NonNull Context context) {
        checkNotNull(context);
        return isRtl(context.getResources());
    }

    @RequiresApi(17)
    @TargetApi(17)
    public static boolean isRtl(@NonNull Resources res) {
        checkNotNull(res);
        return isRtl(res.getConfiguration());
    }

    @RequiresApi(17)
    @TargetApi(17)
    public static boolean isRtl(@NonNull Configuration config) {
        checkNotNull(config);
        return (config.getLayoutDirection() & Configuration.SCREENLAYOUT_LAYOUTDIR_MASK) == Configuration.SCREENLAYOUT_LAYOUTDIR_RTL;
    }

    public static boolean hasTouchscreen(@NonNull Context context) {
        checkNotNull(context);
        return hasTouchscreen(context.getResources());
    }

    public static boolean hasTouchscreen(@NonNull Resources res) {
        checkNotNull(res);
        return hasTouchscreen(res.getConfiguration());
    }

    public static boolean hasTouchscreen(@NonNull Configuration config) {
        checkNotNull(config);
        return config.touchscreen == Configuration.TOUCHSCREEN_FINGER;
    }

    public static boolean hasKeyboard(@NonNull Context context) {
        checkNotNull(context);
        return hasKeyboard(context.getResources());
    }

    public static boolean hasKeyboard(@NonNull Resources res) {
        checkNotNull(res);
        return hasKeyboard(res.getConfiguration());
    }

    public static boolean hasKeyboard(@NonNull Configuration config) {
        checkNotNull(config);
        return config.keyboard == Configuration.KEYBOARD_QWERTY;
    }

    public static boolean isKeyboardHidden(@NonNull Context context) {
        checkNotNull(context);
        return isKeyboardHidden(context.getResources());
    }

    public static boolean isKeyboardHidden(@NonNull Resources res) {
        checkNotNull(res);
        return isKeyboardHidden(res.getConfiguration());
    }

    public static boolean isKeyboardHidden(@NonNull Configuration config) {
        checkNotNull(config);
        return config.keyboardHidden == Configuration.KEYBOARDHIDDEN_YES;
    }

    public static boolean isHardKeyboardHidden(@NonNull Context context) {
        checkNotNull(context);
        return isHardKeyboardHidden(context.getResources());
    }

    public static boolean isHardKeyboardHidden(@NonNull Resources res) {
        checkNotNull(res);
        return isHardKeyboardHidden(res.getConfiguration());
    }

    public static boolean isHardKeyboardHidden(@NonNull Configuration config) {
        checkNotNull(config);
        return config.hardKeyboardHidden == Configuration.HARDKEYBOARDHIDDEN_YES;
    }

    public static boolean hasNavigation(@NonNull Context context) {
        checkNotNull(context);
        return hasNavigation(context.getResources());
    }

    public static boolean hasNavigation(@NonNull Resources res) {
        checkNotNull(res);
        return hasNavigation(res.getConfiguration());
    }

    public static boolean hasNavigation(@NonNull Configuration config) {
        checkNotNull(config);
        return config.navigation != Configuration.NAVIGATION_UNDEFINED && config.navigation != Configuration.NAVIGATION_NONAV;
    }

    public static boolean isNavigationHidden(@NonNull Context context) {
        checkNotNull(context);
        return isNavigationHidden(context.getResources());
    }

    public static boolean isNavigationHidden(@NonNull Resources res) {
        checkNotNull(res);
        return isNavigationHidden(res.getConfiguration());
    }

    public static boolean isNavigationHidden(@NonNull Configuration config) {
        checkNotNull(config);
        return config.navigationHidden == Configuration.NAVIGATIONHIDDEN_YES;
    }

    public static boolean isLandscape(@NonNull Context context) {
        checkNotNull(context);
        return isLandscape(context.getResources());
    }

    public static boolean isLandscape(@NonNull Resources res) {
        checkNotNull(res);
        return isLandscape(res.getConfiguration());
    }

    public static boolean isLandscape(@NonNull Configuration config) {
        checkNotNull(config);
        return config.orientation == Configuration.ORIENTATION_LANDSCAPE;
    }

    @ModeType
    public static int getTypeMode(@NonNull Context context) {
        checkNotNull(context);
        return getTypeMode(context.getResources());
    }

    @ModeType
    public static int getTypeMode(@NonNull Resources res) {
        checkNotNull(res);
        return getTypeMode(res.getConfiguration());
    }

    @ModeType
    public static int getTypeMode(@NonNull Configuration config) {
        checkNotNull(config);
        @ModeType int cur = config.uiMode & Configuration.UI_MODE_TYPE_MASK;
        return cur;
    }

    public static boolean isTypeModeNormal(@NonNull Context context) {
        checkNotNull(context);
        return isTypeModeNormal(context.getResources());
    }

    public static boolean isTypeModeNormal(@NonNull Resources res) {
        checkNotNull(res);
        return isTypeModeNormal(res.getConfiguration());
    }

    public static boolean isTypeModeNormal(@NonNull Configuration config) {
        checkNotNull(config);
        return getTypeMode(config) == Configuration.UI_MODE_TYPE_NORMAL;
    }

    @ModeNight
    public static int getNightMode(@NonNull Context context) {
        checkNotNull(context);
        return getNightMode(context.getResources());
    }

    @ModeNight
    public static int getNightMode(@NonNull Resources res) {
        checkNotNull(res);
        return getNightMode(res.getConfiguration());
    }

    @ModeNight
    public static int getNightMode(@NonNull Configuration config) {
        checkNotNull(config);
        @ModeNight int cur = config.uiMode & Configuration.UI_MODE_NIGHT_MASK;
        return cur;
    }

    public static boolean isNightMode(@NonNull Context context) {
        checkNotNull(context);
        return isNightMode(context.getResources());
    }

    public static boolean isNightMode(@NonNull Resources res) {
        checkNotNull(res);
        return isNightMode(res.getConfiguration());
    }

    public static boolean isNightMode(@NonNull Configuration config) {
        checkNotNull(config);
        return getNightMode(config) == Configuration.UI_MODE_NIGHT_YES;
    }
}

File: src/com/hendraanggrian/support/utils/content/Configurations.kt
package com.hendraanggrian.support.utils.content

import android.annotation.TargetApi
import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import android.support.annotation.IntDef
import android.support.annotation.RequiresApi

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */

@IntDef(Configuration.SCREENLAYOUT_SIZE_UNDEFINED.toLong(),
        Configuration.SCREENLAYOUT_SIZE_SMALL.toLong(),
        Configuration.SCREENLAYOUT_SIZE_NORMAL.toLong(),
        Configuration.SCREENLAYOUT_SIZE_LARGE.toLong(),
        Configuration.SCREENLAYOUT_SIZE_XLARGE.toLong())
@Retention(AnnotationRetention.SOURCE)
annotation class ScreenSize

@RequiresApi(20)
@TargetApi(20)
@IntDef(Configuration.UI_MODE_TYPE_UNDEFINED.toLong(),
        Configuration.UI_MODE_TYPE_NORMAL.toLong(),
        Configuration.UI_MODE_TYPE_DESK.toLong(),
        Configuration.UI_MODE_TYPE_CAR.toLong(),
        Configuration.UI_MODE_TYPE_TELEVISION.toLong(),
        Configuration.UI_MODE_TYPE_APPLIANCE.toLong(),
        Configuration.UI_MODE_TYPE_WATCH.toLong())
@Retention(AnnotationRetention.SOURCE)
annotation class ModeType

@IntDef(Configuration.UI_MODE_NIGHT_UNDEFINED.toLong(),
        Configuration.UI_MODE_NIGHT_NO.toLong(),
        Configuration.UI_MODE_NIGHT_YES.toLong())
@Retention(AnnotationRetention.SOURCE)
annotation class ModeNight

@ScreenSize fun Context.getScreenSize(): Int = resources.getScreenSize()
@ScreenSize fun Resources.getScreenSize(): Int = configuration.getScreenSize()
@ScreenSize fun Configuration.getScreenSize(): Int = screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK

fun Context.isScreenSizeAtLeast(@ScreenSize size: Int): Boolean = resources.isScreenSizeAtLeast(size)
fun Resources.isScreenSizeAtLeast(@ScreenSize size: Int): Boolean = configuration.isScreenSizeAtLeast(size)
fun Configuration.isScreenSizeAtLeast(@ScreenSize size: Int): Boolean = getScreenSize().let { it != Configuration.SCREENLAYOUT_SIZE_UNDEFINED && it >= size }

fun Context.isScreenLong(): Boolean = resources.isScreenLong()
fun Resources.isScreenLong(): Boolean = configuration.isScreenLong()
fun Configuration.isScreenLong(): Boolean = screenLayout and Configuration.SCREENLAYOUT_LONG_MASK == Configuration.SCREENLAYOUT_LONG_YES

@RequiresApi(17) @TargetApi(17) fun Context.isRtl(): Boolean = resources.isRtl()
@RequiresApi(17) @TargetApi(17) fun Resources.isRtl(): Boolean = configuration.isRtl()
@RequiresApi(17) @TargetApi(17) fun Configuration.isRtl(): Boolean = layoutDirection and Configuration.SCREENLAYOUT_LAYOUTDIR_MASK == Configuration.SCREENLAYOUT_LAYOUTDIR_RTL

fun Context.hasTouchscreen(): Boolean = resources.hasTouchscreen()
fun Resources.hasTouchscreen(): Boolean = configuration.hasTouchscreen()
fun Configuration.hasTouchscreen(): Boolean = touchscreen == Configuration.TOUCHSCREEN_FINGER

fun Context.hasKeyboard(): Boolean = resources.hasKeyboard()
fun Resources.hasKeyboard(): Boolean = configuration.hasKeyboard()
fun Configuration.hasKeyboard(): Boolean = keyboard == Configuration.KEYBOARD_QWERTY

fun Context.isKeyboardHidden(): Boolean = resources.isKeyboardHidden()
fun Resources.isKeyboardHidden(): Boolean = configuration.isKeyboardHidden()
fun Configuration.isKeyboardHidden(): Boolean = keyboardHidden == Configuration.KEYBOARDHIDDEN_YES

fun Context.isHardKeyboardHidden(): Boolean = resources.isHardKeyboardHidden()
fun Resources.isHardKeyboardHidden(): Boolean = configuration.isHardKeyboardHidden()
fun Configuration.isHardKeyboardHidden(): Boolean = hardKeyboardHidden == Configuration.HARDKEYBOARDHIDDEN_YES

fun Context.hasNavigation(): Boolean = resources.hasNavigation()
fun Resources.hasNavigation(): Boolean = configuration.hasNavigation()
fun Configuration.hasNavigation(): Boolean = navigation != Configuration.NAVIGATION_UNDEFINED && navigation != Configuration.NAVIGATION_NONAV

fun Context.isNavigationHidden(): Boolean = resources.isNavigationHidden()
fun Resources.isNavigationHidden(): Boolean = configuration.isNavigationHidden()
fun Configuration.isNavigationHidden(): Boolean = navigationHidden == Configuration.NAVIGATIONHIDDEN_YES

fun Context.isLandscape(): Boolean = resources.isLandscape()
fun Resources.isLandscape(): Boolean = configuration.isLandscape()
fun Configuration.isLandscape(): Boolean = orientation == Configuration.ORIENTATION_LANDSCAPE

@ModeType fun Context.getTypeMode(): Int = resources.getTypeMode()
@ModeType fun Resources.getTypeMode(): Int = configuration.getTypeMode()
@ModeType fun Configuration.getTypeMode(): Int = uiMode and Configuration.UI_MODE_TYPE_MASK

fun Context.isTypeModeNormal(): Boolean = resources.isTypeModeNormal()
fun Resources.isTypeModeNormal(): Boolean = configuration.isTypeModeNormal()
fun Configuration.isTypeModeNormal(): Boolean = getTypeMode() == Configuration.UI_MODE_TYPE_NORMAL

@ModeNight fun Context.getNightMode(): Int = resources.getNightMode()
@ModeNight fun Resources.getNightMode(): Int = configuration.getNightMode()
@ModeNight fun Configuration.getNightMode(): Int = uiMode and Configuration.UI_MODE_NIGHT_MASK

fun Context.isNightMode(): Boolean = resources.isNightMode()
fun Resources.isNightMode(): Boolean = configuration.isNightMode()
fun Configuration.isNightMode(): Boolean = getNightMode() == Configuration.UI_MODE_NIGHT_YES